<a href="https://colab.research.google.com/github/RevatiP98/MyDataScienceWork/blob/main/AdvPA_Imputation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Advanced Predictive Analytics
#Data Preprocessing


In [ ]:
 
#Datasets are JobSalary2.csv and ChemicalProcess.csv

In [ ]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer


In [ ]:
job=pd.read_csv("/content/JobSalary2 (1).csv")
job

S_No  Computer  Marketing  Engineer
0     1      84.0       55.0        75
1     2      90.0       53.0        70
2     3      63.0        NaN        88
3     4       NaN       50.0        77
4     5      63.0       41.0        89
5     6      75.0       54.0        85
6     7       NaN       51.0        96
7     8      77.0       64.0       105

In [ ]:
job.isnull()

S_No  Computer  Marketing  Engineer
0  False     False      False     False
1  False     False      False     False
2  False     False       True     False
3  False      True      False     False
4  False     False      False     False
5  False     False      False     False
6  False      True      False     False
7  False     False      False     False

In [ ]:
job.isnull().any()

S_No         False
Computer      True
Marketing     True
Engineer     False
dtype: bool

In [ ]:
#no of missiong values
job.isnull().sum()

S_No         0
Computer     2
Marketing    1
Engineer     0
dtype: int64

In [ ]:
#column names having missing values
job.columns[job.isnull().any()]

Index(['Computer', 'Marketing'], dtype='object')

In [ ]:
#importing imputer
imputer= SimpleImputer()

In [ ]:
imputer.fit(job) #this will replace the missing values with mean value of respective columns

SimpleImputer()

In [ ]:
imp_job=imputer.transform(job)

In [ ]:
imp_job

array([[  1.        ,  84.        ,  55.        ,  75.        ],
       [  2.        ,  90.        ,  53.        ,  70.        ],
       [  3.        ,  63.        ,  52.57142857,  88.        ],
       [  4.        ,  75.33333333,  50.        ,  77.        ],
       [  5.        ,  63.        ,  41.        ,  89.        ],
       [  6.        ,  75.        ,  54.        ,  85.        ],
       [  7.        ,  75.33333333,  51.        ,  96.        ],
       [  8.        ,  77.        ,  64.        , 105.        ]])

In [ ]:
type(job)

pandas.core.frame.DataFrame

In [ ]:
type(imp_job)

numpy.ndarray

In [ ]:
#we can pass pandas object to sklearn but  mostly all of the time we get output as numpy object i.e. array

In [ ]:
pd_job=pd.DataFrame(imp_job, columns=job.columns)
pd_job

S_No   Computer  Marketing  Engineer
0   1.0  84.000000  55.000000      75.0
1   2.0  90.000000  53.000000      70.0
2   3.0  63.000000  52.571429      88.0
3   4.0  75.333333  50.000000      77.0
4   5.0  63.000000  41.000000      89.0
5   6.0  75.000000  54.000000      85.0
6   7.0  75.333333  51.000000      96.0
7   8.0  77.000000  64.000000     105.0

#Chemical Process Dataset

In [ ]:
chem_df=pd.read_csv("/content/ChemicalProcess.csv")
chem_df

Yield  BiologicalMaterial01  BiologicalMaterial02  BiologicalMaterial03  \
0    38.00                  6.25                 49.58                 56.97   
1    42.44                  8.01                 60.97                 67.48   
2    42.03                  8.01                 60.97                 67.48   
3    41.42                  8.01                 60.97                 67.48   
4    42.49                  7.47                 63.33                 72.25   
..     ...                   ...                   ...                   ...   
171  39.66                  6.71                 56.32                 66.19   
172  39.68                  6.87                 56.74                 66.61   
173  42.23                  7.50                 58.41                 68.30   
174  38.48                  7.53                 58.36                 69.25   
175  39.49                  7.53                 58.36                 69.25   

     BiologicalMaterial04  BiologicalMaterial05  BiologicalMaterial06  \
0                   12.74                 19.51                 43.73   
1                   14.65                 19.36                 53.14   
2                   14.65                 19.36                 53.14   
3                   14.65                 19.36                 53.14   
4                   14.02                 17.91                 54.66   
..                    ...                   ...                   ...   
171                 12.35                 20.02                 50.26   
172                 12.55                 20.18                 50.80   
173                 13.33                 20.81                 52.96   
174                 14.35                 20.57                 51.31   
175                 14.35                 20.57                 51.31   

     BiologicalMaterial07  BiologicalMaterial08  BiologicalMaterial09  ...  \
0                   100.0                 16.66                 11.44  ...   
1                   100.0                 19.04                 12.55  ...   
2                   100.0                 19.04                 12.55  ...   
3                   100.0                 19.04                 12.55  ...   
4                   100.0                 18.22                 12.80  ...   
..                    ...                   ...                   ...  ...   
171                 100.0                 17.54                 12.50  ...   
172                 100.0                 17.48                 12.41  ...   
173                 100.0                 17.23                 12.04  ...   
174                 100.0                 17.87                 12.77  ...   
175                 100.0                 17.87                 12.77  ...   

     ManufacturingProcess36  ManufacturingProcess37  ManufacturingProcess38  \
0                     0.019                     0.5                       3   
1                     0.019                     2.0                       2   
2                     0.018                     0.7                       2   
3                     0.018                     1.2                       2   
4                     0.017                     0.2                       2   
..                      ...                     ...                     ...   
171                     NaN                     2.3                       0   
172                     NaN                     1.0                       0   
173                     NaN                     1.3                       0   
174                     NaN                     2.3                       0   
175                     NaN                     0.9                       0   

     ManufacturingProcess39  ManufacturingProcess40  ManufacturingProcess41  \
0                       7.2                     NaN                     NaN   
1                       7.2                     0.1                    0.15   
2                       7.2                 

In [ ]:
chem_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 176 entries, 0 to 175
Data columns (total 58 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Yield                   176 non-null    float64
 1   BiologicalMaterial01    176 non-null    float64
 2   BiologicalMaterial02    176 non-null    float64
 3   BiologicalMaterial03    176 non-null    float64
 4   BiologicalMaterial04    176 non-null    float64
 5   BiologicalMaterial05    176 non-null    float64
 6   BiologicalMaterial06    176 non-null    float64
 7   BiologicalMaterial07    176 non-null    float64
 8   BiologicalMaterial08    176 non-null    float64
 9   BiologicalMaterial09    176 non-null    float64
 10  BiologicalMaterial10    176 non-null    float64
 11  BiologicalMaterial11    176 non-null    float64
 12  BiologicalMaterial12    176 non-null    float64
 13  ManufacturingProcess01  175 non-null    float64
 14  ManufacturingProcess02  173 non-null    fl

In [ ]:
chem_df.isnull().sum()

Yield                      0
BiologicalMaterial01       0
BiologicalMaterial02       0
BiologicalMaterial03       0
BiologicalMaterial04       0
BiologicalMaterial05       0
BiologicalMaterial06       0
BiologicalMaterial07       0
BiologicalMaterial08       0
BiologicalMaterial09       0
BiologicalMaterial10       0
BiologicalMaterial11       0
BiologicalMaterial12       0
ManufacturingProcess01     1
ManufacturingProcess02     3
ManufacturingProcess03    15
ManufacturingProcess04     1
ManufacturingProcess05     1
ManufacturingProcess06     2
ManufacturingProcess07     1
ManufacturingProcess08     1
ManufacturingProcess09     0
ManufacturingProcess10     9
ManufacturingProcess11    10
ManufacturingProcess12     1
ManufacturingProcess13     0
ManufacturingProcess14     1
ManufacturingProcess15     0
ManufacturingProcess16     0
ManufacturingProcess17     0
ManufacturingProcess18     0
ManufacturingProcess19     0
ManufacturingProcess20     0
ManufacturingProcess21     0
ManufacturingP

In [ ]:
chem_df.columns[chem_df.isnull().any()]

Index(['ManufacturingProcess01', 'ManufacturingProcess02',
       'ManufacturingProcess03', 'ManufacturingProcess04',
       'ManufacturingProcess05', 'ManufacturingProcess06',
       'ManufacturingProcess07', 'ManufacturingProcess08',
       'ManufacturingProcess10', 'ManufacturingProcess11',
       'ManufacturingProcess12', 'ManufacturingProcess14',
       'ManufacturingProcess22', 'ManufacturingProcess23',
       'ManufacturingProcess24', 'ManufacturingProcess25',
       'ManufacturingProcess26', 'ManufacturingProcess27',
       'ManufacturingProcess28', 'ManufacturingProcess29',
       'ManufacturingProcess30', 'ManufacturingProcess31',
       'ManufacturingProcess33', 'ManufacturingProcess34',
       'ManufacturingProcess35', 'ManufacturingProcess36',
       'ManufacturingProcess40', 'ManufacturingProcess41'],
      dtype='object')

In [ ]:
from statistics import mean, mode, median

In [ ]:
imputer1=SimpleImputer(strategy='mean')

In [ ]:
imputer1.fit(chem_df)

SimpleImputer()

In [ ]:
imp_chem=imputer1.transform(chem_df)
imp_chem

array([[38.  ,  6.25, 49.58, ...,  3.  ,  1.8 ,  2.4 ],
       [42.44,  8.01, 60.97, ...,  0.9 ,  1.9 ,  2.2 ],
       [42.03,  8.01, 60.97, ...,  1.  ,  1.8 ,  2.3 ],
       ...,
       [42.23,  7.5 , 58.41, ...,  0.6 ,  0.  ,  0.  ],
       [38.48,  7.53, 58.36, ...,  0.5 ,  0.  ,  0.  ],
       [39.49,  7.53, 58.36, ...,  0.6 ,  0.  ,  0.  ]])

In [ ]:
pd_chem=pd.DataFrame(imp_chem, columns=chem_df.columns)
pd_chem

Yield  BiologicalMaterial01  BiologicalMaterial02  BiologicalMaterial03  \
0    38.00                  6.25                 49.58                 56.97   
1    42.44                  8.01                 60.97                 67.48   
2    42.03                  8.01                 60.97                 67.48   
3    41.42                  8.01                 60.97                 67.48   
4    42.49                  7.47                 63.33                 72.25   
..     ...                   ...                   ...                   ...   
171  39.66                  6.71                 56.32                 66.19   
172  39.68                  6.87                 56.74                 66.61   
173  42.23                  7.50                 58.41                 68.30   
174  38.48                  7.53                 58.36                 69.25   
175  39.49                  7.53                 58.36                 69.25   

     BiologicalMaterial04  BiologicalMaterial05  BiologicalMaterial06  \
0                   12.74                 19.51                 43.73   
1                   14.65                 19.36                 53.14   
2                   14.65                 19.36                 53.14   
3                   14.65                 19.36                 53.14   
4                   14.02                 17.91                 54.66   
..                    ...                   ...                   ...   
171                 12.35                 20.02                 50.26   
172                 12.55                 20.18                 50.80   
173                 13.33                 20.81                 52.96   
174                 14.35                 20.57                 51.31   
175                 14.35                 20.57                 51.31   

     BiologicalMaterial07  BiologicalMaterial08  BiologicalMaterial09  ...  \
0                   100.0                 16.66                 11.44  ...   
1                   100.0                 19.04                 12.55  ...   
2                   100.0                 19.04                 12.55  ...   
3                   100.0                 19.04                 12.55  ...   
4                   100.0                 18.22                 12.80  ...   
..                    ...                   ...                   ...  ...   
171                 100.0                 17.54                 12.50  ...   
172                 100.0                 17.48                 12.41  ...   
173                 100.0                 17.23                 12.04  ...   
174                 100.0                 17.87                 12.77  ...   
175                 100.0                 17.87                 12.77  ...   

     ManufacturingProcess36  ManufacturingProcess37  ManufacturingProcess38  \
0                  0.019000                     0.5                     3.0   
1                  0.019000                     2.0                     2.0   
2                  0.018000                     0.7                     2.0   
3                  0.018000                     1.2                     2.0   
4                  0.017000                     0.2                     2.0   
..                      ...                     ...                     ...   
171                0.019573                     2.3                     0.0   
172                0.019573                     1.0                     0.0   
173                0.019573                     1.3                     0.0   
174                0.019573                     2.3                     0.0   
175                0.019573                     0.9                     0.0   

     ManufacturingProcess39  ManufacturingProcess40  ManufacturingProcess41  \
0                       7.2                0.017714                0.023714   
1                       7.2                0.100000                0.150000   
2                       7.2                0

In [ ]:
pd_chem.columns[pd_chem.isnull().any()]

Index([], dtype='object')

In [ ]:
pd_chem.columns

Index(['Yield', 'BiologicalMaterial01', 'BiologicalMaterial02',
       'BiologicalMaterial03', 'BiologicalMaterial04', 'BiologicalMaterial05',
       'BiologicalMaterial06', 'BiologicalMaterial07', 'BiologicalMaterial08',
       'BiologicalMaterial09', 'BiologicalMaterial10', 'BiologicalMaterial11',
       'BiologicalMaterial12', 'ManufacturingProcess01',
       'ManufacturingProcess02', 'ManufacturingProcess03',
       'ManufacturingProcess04', 'ManufacturingProcess05',
       'ManufacturingProcess06', 'ManufacturingProcess07',
       'ManufacturingProcess08', 'ManufacturingProcess09',
       'ManufacturingProcess10', 'ManufacturingProcess11',
       'ManufacturingProcess12', 'ManufacturingProcess13',
       'ManufacturingProcess14', 'ManufacturingProcess15',
       'ManufacturingProcess16', 'ManufacturingProcess17',
       'ManufacturingProcess18', 'ManufacturingProcess19',
       'ManufacturingProcess20', 'ManufacturingProcess21',
       'ManufacturingProcess22', 'ManufacturingProce

In [ ]:
X=pd_chem.drop('Yield',axis=1)

In [ ]:
Y=pd_chem['Yield']

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test=train_test_split(X,Y, test_size=0.3, random_state=2022)

In [ ]:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()

In [ ]:
lr.fit(x_train, y_train)

LinearRegression()

In [ ]:
y_pred=lr.predict(x_test)

In [ ]:
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
print("R2_Score: with 'Mean' as a Strategy :",r2_score(y_test,y_pred))

R2_Score: with 'Mean' as a Strategy : -0.39647934947111874


In [ ]:
imputer2=SimpleImputer(strategy='median')

In [ ]:
imputer2.fit(chem_df)

SimpleImputer(strategy='median')

In [ ]:
imp_chem1=imputer2.transform(chem_df)
imp_chem1

array([[38.  ,  6.25, 49.58, ...,  3.  ,  1.8 ,  2.4 ],
       [42.44,  8.01, 60.97, ...,  0.9 ,  1.9 ,  2.2 ],
       [42.03,  8.01, 60.97, ...,  1.  ,  1.8 ,  2.3 ],
       ...,
       [42.23,  7.5 , 58.41, ...,  0.6 ,  0.  ,  0.  ],
       [38.48,  7.53, 58.36, ...,  0.5 ,  0.  ,  0.  ],
       [39.49,  7.53, 58.36, ...,  0.6 ,  0.  ,  0.  ]])

In [ ]:
pd_chem1=pd.DataFrame(imp_chem1, columns=chem_df.columns)
pd_chem1

Yield  BiologicalMaterial01  BiologicalMaterial02  BiologicalMaterial03  \
0    38.00                  6.25                 49.58                 56.97   
1    42.44                  8.01                 60.97                 67.48   
2    42.03                  8.01                 60.97                 67.48   
3    41.42                  8.01                 60.97                 67.48   
4    42.49                  7.47                 63.33                 72.25   
..     ...                   ...                   ...                   ...   
171  39.66                  6.71                 56.32                 66.19   
172  39.68                  6.87                 56.74                 66.61   
173  42.23                  7.50                 58.41                 68.30   
174  38.48                  7.53                 58.36                 69.25   
175  39.49                  7.53                 58.36                 69.25   

     BiologicalMaterial04  BiologicalMaterial05  BiologicalMaterial06  \
0                   12.74                 19.51                 43.73   
1                   14.65                 19.36                 53.14   
2                   14.65                 19.36                 53.14   
3                   14.65                 19.36                 53.14   
4                   14.02                 17.91                 54.66   
..                    ...                   ...                   ...   
171                 12.35                 20.02                 50.26   
172                 12.55                 20.18                 50.80   
173                 13.33                 20.81                 52.96   
174                 14.35                 20.57                 51.31   
175                 14.35                 20.57                 51.31   

     BiologicalMaterial07  BiologicalMaterial08  BiologicalMaterial09  ...  \
0                   100.0                 16.66                 11.44  ...   
1                   100.0                 19.04                 12.55  ...   
2                   100.0                 19.04                 12.55  ...   
3                   100.0                 19.04                 12.55  ...   
4                   100.0                 18.22                 12.80  ...   
..                    ...                   ...                   ...  ...   
171                 100.0                 17.54                 12.50  ...   
172                 100.0                 17.48                 12.41  ...   
173                 100.0                 17.23                 12.04  ...   
174                 100.0                 17.87                 12.77  ...   
175                 100.0                 17.87                 12.77  ...   

     ManufacturingProcess36  ManufacturingProcess37  ManufacturingProcess38  \
0                     0.019                     0.5                     3.0   
1                     0.019                     2.0                     2.0   
2                     0.018                     0.7                     2.0   
3                     0.018                     1.2                     2.0   
4                     0.017                     0.2                     2.0   
..                      ...                     ...                     ...   
171                   0.020                     2.3                     0.0   
172                   0.020                     1.0                     0.0   
173                   0.020                     1.3                     0.0   
174                   0.020                     2.3                     0.0   
175                   0.020                     0.9                     0.0   

     ManufacturingProcess39  ManufacturingProcess40  ManufacturingProcess41  \
0                       7.2                     0.0                    0.00   
1                       7.2                     0.1                    0.15   
2                       7.2                 

In [ ]:
pd_chem1.columns[pd_chem1.isnull().any()]

Index([], dtype='object')

In [ ]:
X1=pd_chem1.drop('Yield',axis=1)

In [ ]:
Y1=pd_chem1['Yield']

In [ ]:
x_train1, x_test1, y_train1, y_test1=train_test_split(X1,Y1, test_size=0.3, random_state=2022)

In [ ]:
lr1=LinearRegression()

In [ ]:
lr1.fit(x_train1, y_train1)

LinearRegression()

In [ ]:
y_pred1=lr1.predict(x_test1)

In [ ]:
print("R2_Score: with 'Median' as a Strategy :",r2_score(y_test1,y_pred1+
                                                         ))


R2_Score: with 'Median' as a Strategy : -0.3700595666117459


#So as we can see when startegy= median we got r2_score -0.37 which is greater than r2_score -0.39  when startegy = mean hence we will consider stategy as Median.
